ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java

```diff
@@ -405,6 +406,42 @@ void updatePartitions(Collection<? extends PartitionId> allPartitions, String lo
       return healthyWritablePartitions.isEmpty() ? writablePartitions : healthyWritablePartitions;
     }
 
+    /**
+     * Returns a writable partition selected at random, that belongs to the specified partition class and that is in the
+     * state {@link PartitionState#READ_WRITE} AND has all replicas up. In case none of the partitions have all replicas
+     * up, any writable partition is returned.
+     * @param partitionClass the class of the partitions desired. Can be {@code null}.
+     * @param partitionsToExclude partitions that should be excluded from the result. Can be {@code null} or empty.
+     * @return A writable partition or {@code null} if no writable partition with given criteria found
+     */
+    PartitionId getRandomWritablePartition(String partitionClass, List<PartitionId> partitionsToExclude) {
+      PartitionId selectedPartition = null;
+      PartitionId anyWritablePartition = null;
+      List<PartitionId> partitionsInClass = getPartitionsInClass(partitionClass, true);
+
+      int workingSize = partitionsInClass.size();
+      while(true) {
+        if(workingSize == 0)
+          break;
+        int randomIndex = ThreadLocalRandom.current().nextInt(workingSize);
+        PartitionId selected = partitionsInClass.get(randomIndex);
+        if(partitionsToExclude == null || partitionsToExclude.size() == 0 || !partitionsToExclude.contains(selected)) {
+          if (selected.getPartitionState() == PartitionState.READ_WRITE) {
+            anyWritablePartition = selected;
+            if(areAllReplicasForPartitionUp(selected)) {
```

Comment:
I remember during offline discussion, you even proposed to only check if local replicas are up. Will we try that in the future PR?

Reply:
fixed,